In [1]:
import pandas as pd
import statsmodels.api as sm

# Premier traitement des données

In [2]:
data = pd.read_csv("Titanic.csv")

In [3]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
data_clean = data.drop(columns = ["PassengerId", "Ticket", "Cabin", "Embarked"])
data_clean['Age'] =  data_clean['Age'].fillna(value = data_clean['Age'].mean())

In [5]:
print(data_clean.shape)
data_clean.head()

(891, 8)


Survived  Pclass                                               Name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      Sex   Age  SibSp  Parch     Fare  
0    male  22.0      1      0   7.2500  
1  female  38.0      1      0  71.2833  
2  female  26.0      0      0   7.9250  
3  female  35.0      1      0  53.1000  
4    male  35.0      0      0   8.0500

## Exercice 1

In [6]:
alive = data_clean['Survived'].sum()
total = data_clean['Survived'].shape[0]

print(f"{total - alive} passagers sont décédés, ce qui répresente {(100 * (total - alive) / total):.2f} % du total.")
print()


549 passagers sont décédés, ce qui répresente 61.62 % du total.



## Exercice 2

In [7]:
total_man = data_clean.groupby('Sex')['Name'].count()['male']
total_woman = data_clean.groupby('Sex')['Name'].count()['female']

print(f"Percentage d'hommes : {(100*total_man/(total_woman+total_man)):.2f} %, percentage de femmes {(100*total_woman/(total_woman+total_man)):.2f} %.")

Percentage d'hommes : 64.76 %, percentage de femmes 35.24 %.


In [8]:
data_clean_dead = data_clean.loc[data_clean['Survived'] == 0]

dead_man = data_clean_dead.groupby('Sex')['Name'].count()['male']
dead_woman = data_clean_dead.groupby('Sex')['Name'].count()['female']

print(f"Parmi les personnes décédés, {(dead_man/(dead_man + dead_woman)*100):.2f} % ont été des hommes et {(dead_woman/(dead_man + dead_woman)*100):.2f} % des femmes.")

Parmi les personnes décédés, 85.25 % ont été des hommes et 14.75 % des femmes.


In [9]:
data_clean_alive = data_clean.loc[data_clean['Survived'] == 1]

alive_man = data_clean_alive.groupby('Sex')['Name'].count()['male']
alive_woman = data_clean_alive.groupby('Sex')['Name'].count()['female']

print(f"Parmi les personnes survivantes, {(alive_man/(alive_man + alive_woman)*100):.2f} % ont été des hommes et {(alive_woman/(alive_man + alive_woman)*100):.2f} % des femmes.")

Parmi les personnes survivantes, 31.87 % ont été des hommes et 68.13 % des femmes.


Observation : Malgré le fait que la majorité de la population du titanic était composé des hommes, on observe que la plupart des personnes qui ont survecu l'accident ont été des femmes (68.13 % contre 31.87 %) et plus d'hommes ont décédé (85.25 % contre 14.75%).

## Exercice 3

In [10]:
n_first = data_clean['Pclass'].value_counts()[1]
n_second = data_clean['Pclass'].value_counts()[2]
n_third = data_clean['Pclass'].value_counts()[3]

n_total = data_clean['Pclass'].count()

print(f"Il y avait {n_first} passagers en première classe ({(100*n_first/n_total):.2f} %), {n_second} en deuxième ({(100*n_second/n_total):.2f} %) et {n_third} en troisième ({(100*n_third/n_total):.2f} %).")

Il y avait 216 passagers en première classe (24.24 %), 184 en deuxième (20.65 %) et 491 en troisième (55.11 %).


In [11]:
total_dead = dead_man + dead_woman

dead_first = data_clean_dead.groupby('Pclass')['Name'].count()[1]
dead_second = data_clean_dead.groupby('Pclass')['Name'].count()[2]
dead_third = data_clean_dead.groupby('Pclass')['Name'].count()[3]

print("Au total : \n")

print(f"Parmi les personnes décédés, {(dead_first/(total_dead)*100):.2f} % ont été de première classe.")
print(f"Parmi les personnes décédés, {(dead_second/(total_dead)*100):.2f} % ont été de deuxième classe.")
print(f"Parmi les personnes décédés, {(dead_third/(total_dead)*100):.2f} % ont été de troisième classe.\n")

print("On regarde dans une même classe : \n")
print(f"Parmi les personnes de première classe, {(100*dead_first / n_first):.2f} % sont décédés")
print(f"Parmi les personnes de deuxième classe, {(100*dead_second / n_second):.2f} % sont décédés")
print(f"Parmi les personnes de troisième classe, {(100*dead_third / n_third):.2f} % sont décédés")

Au total : 

Parmi les personnes décédés, 14.57 % ont été de première classe.
Parmi les personnes décédés, 17.67 % ont été de deuxième classe.
Parmi les personnes décédés, 67.76 % ont été de troisième classe.

On regarde dans une même classe : 

Parmi les personnes de première classe, 37.04 % sont décédés
Parmi les personnes de deuxième classe, 52.72 % sont décédés
Parmi les personnes de troisième classe, 75.76 % sont décédés


In [12]:
total_alive = alive_man + alive_woman

alive_first = data_clean_alive.groupby('Pclass')['Name'].count()[1]
alive_second = data_clean_alive.groupby('Pclass')['Name'].count()[2]
alive_third = data_clean_alive.groupby('Pclass')['Name'].count()[3]

print("Au total : \n")

print(f"Parmi les personnes qui ont survécu, {(alive_first/(total_alive)*100):.2f} % ont été de première classe.")
print(f"Parmi les personnes qui ont survécu, {(alive_second/(total_alive)*100):.2f} % ont été de deuxième classe.")
print(f"Parmi les personnes qui ont survécu, {(alive_third/(total_alive)*100):.2f} % ont été de troisième classe.\n")

print("On regarde dans une même classe : \n")
print(f"Parmi les personnes de première classe, {(100*alive_first / n_first):.2f} % ont survécu")
print(f"Parmi les personnes de deuxième classe, {(100*alive_second / n_second):.2f} % ont survécu")
print(f"Parmi les personnes de troisième classe, {(100*alive_third / n_third):.2f} % ont survécu")

Au total : 

Parmi les personnes qui ont survécu, 39.77 % ont été de première classe.
Parmi les personnes qui ont survécu, 25.44 % ont été de deuxième classe.
Parmi les personnes qui ont survécu, 34.80 % ont été de troisième classe.

On regarde dans une même classe : 

Parmi les personnes de première classe, 62.96 % ont survécu
Parmi les personnes de deuxième classe, 47.28 % ont survécu
Parmi les personnes de troisième classe, 24.24 % ont survécu


Observations : On observe que la plupart des personnes qui ont survecu ont été dans la première classe (aussi plus grad taux de survie). En plus ce sont ceux qui étaient dans la troisième classe qui sont décédés le plus (aussi plus grand taux de mort).

## Exercice 4

In [13]:
data_clean_corr = data_clean.copy()

# Turns Sex into numerical
data_clean_corr['Sex'] = data_clean['Sex'].map(lambda x: 1 if x == 'male' else 0)

# Creates category columns for Class
data_clean_corr = data_clean_corr.join(pd.get_dummies(data_clean_corr['Pclass'])).drop(columns='Pclass')
data_clean_corr = data_clean_corr.rename(columns={1: '1st Class', 2: '2nd Class', 3: '3rd Class'})

# Covariables
covar_names = ['Sex', 'Age', 'SibSp', 'Parch', 'Fare', '1st Class', '2nd Class', '3rd Class']

display(data_clean_corr.corr())

Survived       Sex       Age     SibSp     Parch      Fare  \
Survived   1.000000 -0.543351 -0.069809 -0.035322  0.081629  0.257307   
Sex       -0.543351  1.000000  0.084153 -0.114631 -0.245489 -0.182333   
Age       -0.069809  0.084153  1.000000 -0.232625 -0.179191  0.091566   
SibSp     -0.035322 -0.114631 -0.232625  1.000000  0.414838  0.159651   
Parch      0.081629 -0.245489 -0.179191  0.414838  1.000000  0.216225   
Fare       0.257307 -0.182333  0.091566  0.159651  0.216225  1.000000   
1st Class  0.285904 -0.098013  0.319916 -0.054582 -0.017633  0.591711   
2nd Class  0.093349 -0.064746  0.006589 -0.055932 -0.000734 -0.118557   
3rd Class -0.322308  0.137143 -0.281004  0.092548  0.015790 -0.413333   

           1st Class  2nd Class  3rd Class  
Survived    0.285904   0.093349  -0.322308  
Sex        -0.098013  -0.064746   0.137143  
Age         0.319916   0.006589  -0.281004  
SibSp      -0.054582  -0.055932   0.092548  
Parch      -0.017633  -0.000734   0.015790  
Fare        0.591711  -0.118557  -0.413333  
1st Class   1.000000  -0.288585  -0.626738  
2nd Class  -0.288585   1.000000  -0.565210  
3rd Class  -0.626738  -0.565210   1.000000

 Nous observons que certaines variables sont fortement corrélées avec le taux de survie, comme, par exemple, le sexe du passager et la classe dans laquele il voyage. Cela suggère que ces variables peuvent être utilisées pour prédire la probabilité de survie d’un passager. De plus, certaines variables sont corrélées entre elles, comme le tarif du billet et la classe (un billet de meilleure classe coûte plus). 

## Exercice 5

In [14]:
alpha=.05
Xs = data_clean_corr[covar_names]

logit_reg = sm.Logit(data_clean_corr[['Survived']], Xs).fit()
print(logit_reg.summary())

Optimization terminated successfully.
         Current function value: 0.442576
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:               Survived   No. Observations:                  891
Model:                          Logit   Df Residuals:                      883
Method:                           MLE   Df Model:                            7
Date:                Wed, 19 Oct 2022   Pseudo R-squ.:                  0.3354
Time:                        23:13:30   Log-Likelihood:                -394.34
converged:                       True   LL-Null:                       -593.33
Covariance Type:            nonrobust   LLR p-value:                 6.452e-82
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Sex           -2.7609      0.199    -13.856      0.000      -3.151      -2.370
Age           -0.0395      0.

In [15]:
logit_reg.conf_int(alpha).rename({0: '5%', 1: '95%'}, axis=1)

5%       95%
Sex       -3.151458 -2.370395
Age       -0.054863 -0.024117
SibSp     -0.564915 -0.135246
Parch     -0.343866  0.117198
Fare      -0.001804  0.007786
1st Class  2.965671  4.716044
2nd Class  2.135153  3.500336
3rd Class  1.121332  2.260654

Les intervalles de confiance montrent que les variables ‘Parch’ et ‘Fare’ ne sont pas significatives à niveau 5% car ils contiennent zéro.

## Exercice 6

In [16]:
print(logit_reg.pvalues)

Sex          1.165424e-43
Age          4.781978e-07
SibSp        1.403955e-03
Parch        3.352689e-01
Fare         2.214914e-01
1st Class    7.865927e-18
2nd Class    5.929875e-16
3rd Class    5.955611e-09
dtype: float64


Le p-valeur obtenu pour le test sur ‘Parch’ confirme l’observation faite dans l’exercice 5. La haute p-valeur indique que, pour notre modèle, cette variable ne porte pas beaucoup d’information sur le taux de survie.

## Exercice 7

In [17]:
covars_noparch = covar_names.copy()
covars_noparch.remove('Parch')

Xs = data_clean_corr[covars_noparch]

logit_reg = sm.Logit(data_clean_corr[['Survived']], Xs).fit()
print(logit_reg.summary())

Optimization terminated successfully.
         Current function value: 0.443106
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:               Survived   No. Observations:                  891
Model:                          Logit   Df Residuals:                      884
Method:                           MLE   Df Model:                            6
Date:                Wed, 19 Oct 2022   Pseudo R-squ.:                  0.3346
Time:                        23:13:30   Log-Likelihood:                -394.81
converged:                       True   LL-Null:                       -593.33
Covariance Type:            nonrobust   LLR p-value:                 1.210e-82
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Sex           -2.7245      0.195    -13.981      0.000      -3.106      -2.343
Age           -0.0392      0.

In [18]:
logit_reg.conf_int(alpha).rename({0: '5%', 1: '95%'}, axis=1)

5%       95%
Sex       -3.106472 -2.342584
Age       -0.054544 -0.023879
SibSp     -0.586601 -0.170082
Fare      -0.002077  0.007030
1st Class  2.945675  4.685900
2nd Class  2.093773  3.439090
3rd Class  1.077795  2.190121

Les nouveaux intervalles de confiance indiquent que la variable ‘Fare’ n’est toujours pas significative à niveau 5%.

## Exercice 8

In [19]:
print(logit_reg.pvalues)

Sex          2.034713e-44
Age          5.371748e-07
SibSp        3.699590e-04
Fare         2.864641e-01
1st Class    8.310466e-18
2nd Class    7.585026e-16
3rd Class    8.501479e-09
dtype: float64


La p-valeur montre, comme pour ‘Parch’, que ‘Fare’ n’est pas une variable significative pour notre modèle. Cela peut s’expliquer par la forte corrélation entre cette variable et la classe du passager, rendant son utilisation redondante quand la classe est déjà fournie au modèle.

## Exercice 9

In [20]:
covars_nofare = covars_noparch.copy()
covars_nofare.remove('Fare')

Xs = data_clean_corr[covars_nofare]

logit_reg = sm.Logit(data_clean_corr[['Survived']], Xs).fit()
print(logit_reg.summary())

Optimization terminated successfully.
         Current function value: 0.443793
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:               Survived   No. Observations:                  891
Model:                          Logit   Df Residuals:                      885
Method:                           MLE   Df Model:                            5
Date:                Wed, 19 Oct 2022   Pseudo R-squ.:                  0.3336
Time:                        23:13:30   Log-Likelihood:                -395.42
converged:                       True   LL-Null:                       -593.33
Covariance Type:            nonrobust   LLR p-value:                 2.366e-83
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Sex           -2.7402      0.194    -14.110      0.000      -3.121      -2.360
Age           -0.0399      0.

In [21]:
logit_reg.conf_int(alpha).rename({0: '5%', 1: '95%'}, axis=1)

5%       95%
Sex       -3.120844 -2.359592
Age       -0.055136 -0.024570
SibSp     -0.562530 -0.153989
1st Class  3.243699  4.811148
2nd Class  2.176266  3.498922
3rd Class  1.128785  2.230466

In [22]:
print(logit_reg.pvalues)

Sex          3.284395e-45
Age          3.206039e-07
SibSp        5.871505e-04
1st Class    7.353226e-24
2nd Class    4.110044e-17
3rd Class    2.282220e-09
dtype: float64


Toutes les variables considérées dans le nouveau modèle ont des coefficients non nuls à niveau 5%. Indiquant que toutes les variables sont importantes pour la prévision du taux de survie.

## Exercice 10

In [23]:
ex_man = pd.DataFrame(data={'Sex': [1], 'Age': [22], 'SibSp': [0], '1st Class': [1], '2nd Class': [0], '3rd Class': [0]})
ex_woman = pd.DataFrame(data={'Sex': [0], 'Age': [22], 'SibSp': [0], '1st Class': [1], '2nd Class': [0], '3rd Class': [0]})

p_man = logit_reg.predict(ex_man)[0] 
p_woman = logit_reg.predict(ex_woman)[0]

print("Le modele nous dit que:")
print(f"La probabilité de survie de l'homme décrit par la question 10. est de {p_man * 100:.2f} %")
print(f"La probabilité de survie de la femme est de {p_woman * 100:.2f} %")

Le modele nous dit que:
La probabilité de survie de l'homme décrit par la question 10. est de 60.12 %
La probabilité de survie de la femme est de 95.89 %
